### PR TITLE
feat: add support for KEMs

### PR DESCRIPTION
**Pull request by TWal**
This PR adds support for KEMs.

Putting this as a draft, as this needs:
- some comments
- to be tested in protocol proof, to make sure that it is actually usable

I will take care of that later!

### CODE REVIEW

**Comment by TWal** [comment]
I added the comments and tested it in a protocol proof, I think this is ready to go. Asking @qaphla or @fabian-hk for review because I think you have been looking at the cryptographic invariants recently.

File: src/core/DY.Core.Bytes.fst [hunk 1]
```diff
@@ -1916,3 +1990,232 @@ val get_label_kdf_expand:
...
+  sk:bytes -> encap:bytes ->
+  Lemma
+  (requires
+    bytes_invariant tr sk /\
+    bytes_invariant tr encap /\
+    (KemKey? (get_usage sk) \/ (get_label sk) `can_flow tr` public)
+  )
+  (ensures (
+    match kem_decap sk encap with
+    | Some ss -> get_label ss `can_flow tr` public \/ get_usage sk == KemKey (get_usage ss)
```
Comment (fabian-hk) [inline src/core/DY.Core.Bytes.fst:2176]: This is only a small thing, but in my opinion, the ensures would be better readable if we invert the order of the OR condition so that it says either it has this usage or it can flow to public.

File: src/core/DY.Core.Bytes.fst [hunk 2]
```diff
@@ -329,6 +346,26 @@ let get_dh_usage #cusages pk =
   | DhPub sk -> get_usage sk
   | _ -> NoUsage
 
+/// Obtain the label of the corresponding KEM private key of a KEM public key.
+/// Although the KEM public key label is public,
+/// this is useful to reason on the corresponding KEM private key label.
+
+[@@"opaque_to_smt"]
+val get_kem_label: {|crypto_usages|} -> bytes -> GTot label
+let get_kem_label #cusages pk =
```
Comment (fabian-hk) [inline src/core/DY.Core.Bytes.fst:355]: I think we should name this function `get_kem_sk_label` to better describe what the function does.

File: src/core/DY.Core.Bytes.fst [hunk 3]
```diff
@@ -329,6 +346,26 @@ let get_dh_usage #cusages pk =
...
+  match pk with
+  | KemPub sk -> get_label sk
+  | _ -> public
+
+/// Same as above, for usage.
+
+[@@"opaque_to_smt"]
+val get_kem_usage: {|crypto_usages|} -> bytes -> GTot usage
+let get_kem_usage #cusages pk =
```
Comment (fabian-hk) [inline src/core/DY.Core.Bytes.fst:364]: Same as above `get_kem_sk_usage`.

**Comment by fabian-hk** [comment]
I think this PR looks good, as far as I can tell. I just made a few small comments.

File: src/core/DY.Core.Bytes.fst [hunk 4]
```diff
@@ -1270,6 +1342,7 @@ let get_label_pk_enc #cusages pk nonce msg =
 
 /// User lemma (public-key decryption bytes invariant).
 
+#push-options "--z3rlimit 25"
```
Comment (fabian-hk) [inline src/core/DY.Core.Bytes.fst:1345]: Do we need to change the rlimit here? This seems unrelated to the KEM implementation.

Comment (TWal) [reply src/core/DY.Core.Bytes.fst:1345]: Yes we need, I don't really know why, but sometimes the rlimit must be bumped when the SMT gods decide so. Here, it might be because `bytes` has more constructors.

File: src/core/DY.Core.Bytes.fst (hunk 1 above)
Comment (TWal) [reply src/core/DY.Core.Bytes.fst]: Done

File: src/core/DY.Core.Bytes.fst (hunk 2 above)
Comment (TWal) [reply src/core/DY.Core.Bytes.fst]: Done, and we will need to give the same treatment to similar DH functions at some point

File: src/core/DY.Core.Bytes.fst (hunk 3 above)
Comment (TWal) [reply src/core/DY.Core.Bytes.fst]: Done

**Review by fabian-hk — APPROVED** [review]
